### PR TITLE
refactor!: Change CRD for rack awareness according to decision

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

According to https://github.com/stackabletech/decisions/issues/2

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [x] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -1512,7 +1502,9 @@ spec:
       default:
         replicas: 1";
 
-        let hdfs: HdfsCluster = serde_yaml::from_str(cr).unwrap();
+        let deserializer = serde_yaml::Deserializer::from_str(cr);
+        let hdfs: HdfsCluster =
+            serde_yaml::with::singleton_map_recursive::deserialize(deserializer).unwrap();
```
Comment (NickLarsenNZ) [inline rust/crd/src/lib.rs:1508]: TIL (singleton_map and singleton_map_recursive) :D

File: rust/crd/src/lib.rs [hunk 2]
```diff
@@ -184,33 +184,25 @@ pub struct HdfsClusterConfig {
...
-    label_type: TopologyLabelType,
-
-    /// Name of the label that will be used to resolve a datanode to a topology zone.
-    label_name: String,
+pub enum TopologyLabel {
+    /// Name of the label on the Kubernetes Node (where the Pod is placed on) used to resolve a datanode to a topology
+    /// zone.
+    NodeLabel(String),
+    /// Name of the label on the Kubernetes Pod used to resolve a datanode to a topology zone.
+    PodLabel(String),
```
Comment (NickLarsenNZ) [inline rust/crd/src/lib.rs]: code-style: https://docs.stackable.tech/home/nightly/contributor/code-style-guide#_formatting_of_struct_fields_and_enum_variants

Fields or Variants with comments/attributes need empty lines between.

```suggestion
pub enum TopologyLabel {
    /// Name of the label on the Kubernetes Node (where the Pod is placed on) used to resolve a datanode to a topology
    /// zone.
    NodeLabel(String),
    
    /// Name of the label on the Kubernetes Pod used to resolve a datanode to a topology zone.
    PodLabel(String),
```

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM

File: rust/crd/src/lib.rs (hunk 2 above)
Comment (sbernauer) [reply rust/crd/src/lib.rs]: Sure, forgot that :see_no_evil:

**Review by NickLarsenNZ — APPROVED** [review]
LGTM

**Pull request merged**